pgwire: split more code into pre-serve and serve logic

knz opened the pull request:
Parent PRs:
- [x] https://github.com/cockroachdb/cockroach/pull/84608
- [x] https://github.com/cockroachdb/cockroach/pull/91739
- [x] https://github.com/cockroachdb/cockroach/pull/91744
- [x] https://github.com/cockroachdb/cockroach/pull/92574
- [x] https://github.com/cockroachdb/cockroach/pull/92575

This PR splits the pre-connection logic (tenant-independent) into its own code sections.

Epic: CRDB-14537

knz committed:
pgwire: unexport ServerMetrics, rename to tenantSpecificMetrics

Release note: None

knz committed:
pgwire: fix typo in metric descriptions

Release note: None

knz committed:
pgwire: create a framework for pre-serve connection logic

This commit introduces `PreServeConnHandler`, which will
handle the preparation of a SQL connection prior to handing it off to
a specific tenant server.

Release note (ops changes): The count of network bytes sent
to report re-authentication errors to a SQL client is now
reported via the metric `sql.pre_serve.bytesout` (instead of
`sql.bytesout` previously). The count of pre-authentication errors
is now reported via the metric `sql.pre_serve.conn.failures` (instead
of `sql.conn.failures` previously).

Release note (ops changes): The count of new SQL connections
is now also reported on `sql.pre_serve.new_conns`.

knz committed:
pgwire: simplify writeErr/sendErr

Release note: None

knz committed:
pgwire: lift readVersion + TLS handshake into the pre-serve handler

Release note (ops change): The bytes read from SQL clients
prior to authentication are now reported via the metric
`sql.pre_serve.bytesin` (instead of `sql.bytesin` previously).

knz committed:
pgwire: make handleCancel unaware of conn/buf

Release note: None

knz committed:
pgwire: split `parseClientProvidedSessionParameters`

We need the read/parse of the parameters to be tenant-independent,
so the function cannot contain tenant-specific configuration.

Release note: None

knz committed:
pgwire: split the pre-conn reserved memory account

This commit extracts the memory reservation into
a tenant-independent memory pool.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/92576)
<!-- Reviewable:end -->

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/92576)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92576)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[`pkg/sql/pgwire/server.go` line 987 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/92576#-NI3b5NG9RyfxC5ZZOt9:-NI3b5NG9RyfxC5ZZOtA:b-qgwo2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/10cf4c8bb3b2c9cff29597249a42ea6488904930/pkg/sql/pgwire/server.go#L987)):*
> ```Go
> 	// Replace RemoteAddr if specified by client and we trust the client.
> 	if cp.clientProvidedRemoteAddr != "" {
> 		if !trustClientProvidedRemoteAddr {
> ```

since this `trustClientProvidedRemoteAddr` comes from an env var, i don't see why the client-provided address logic is tenant-specfic
___
*[`pkg/sql/pgwire/server.go` line 258 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/92576#-NI3c8Kj2lrrcYXZtToh:-NI3c8Kj2lrrcYXZtToi:bui2qk7) ([raw file](https://github.com/cockroachdb/cockroach/blob/521774a83ddf6ddc43298591d96f78f0812e6c79/pkg/sql/pgwire/server.go#L258)):*
> ```Go
> 	// tenantConnMonitor is the pool where the memory usage for the
> 	// initial connection overhead is accounted for.
> 	tenantConnMonitor *mon.BytesMonitor
> ```

should this be called `tenantSpecificConnMonitor`? it seems like that `tenantSpecificXYZ` convention was used elsewhere


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92576)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @rafiss)

___
*[`pkg/sql/pgwire/server.go` line 987 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/92576#-NI3b5NG9RyfxC5ZZOt9:-NIbWA3n0y6mtNkTjGLx:bp7ay4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/69040924931b05ff46933b4c55c49e632492938b/pkg/sql/pgwire/server.go#L987)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

since this `trustClientProvidedRemoteAddr` comes from an env var, i don't see why the client-provided address logic is tenant-specfic
</blockquote></details>

You are very right. I'm moving it there in #92579.
___
*[`pkg/sql/pgwire/server.go` line 258 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/92576#-NI3c8Kj2lrrcYXZtToh:-NIbW9Ox6H0Q6_Jo6n0w:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/040237785177d2485918a87319829b35fd4f18b1/pkg/sql/pgwire/server.go#L258)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should this be called `tenantSpecificConnMonitor`? it seems like that `tenantSpecificXYZ` convention was used elsewhere
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
> Reviewed 1 of 5 files at r5, 4 of 4 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/92576)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)



<!-- Sent from Reviewable.io -->

knz committed:
cli/demo: fix the port allocation

Prior to this patch, we were unconditionally applying a port offset
for secondary tenant servers, even when the port allocation for the KV
node was randomized. This resulted in the demo cluster trying to
reuse arbitrary TCP ports which may be already in use elsewhere in the
system.

This patch fixes it by ensuring that tenant servers also use random
allocation when the KV node ports were randomly allocated in the first
place.

It also adds demo networking tests that verify this via the
`--sql-port` / `--http-port` flags.

Release note: None

knz committed:
cli/demo: deflake some interactive tests

Some of the tests were not waiting for the prompt to start sending
their input. This was causing the input to be lost in some cases.

Release note: None
